### PR TITLE
MRG: Use ruff-format instead of Black

### LINKED ISSUE #12255: We may be able to ditch Black in favor of just Ruff
Ruff has started offering formatting functionality a couple of months back, in addition to just lining. They claim they tested it on large codebases like Django and it yields >99.9% identical results to Black. Differences are e.g. that it would complain about comments exceeding the max. line length, while Black doesn't.

https://astral.sh/blog/the-ruff-formatter

https://docs.astral.sh/ruff/formatter/

In any case, we could theoretically drop Black from our CI and pre-commit hooks and simply use Ruff for everything. This would speed up the pre-commit checks.

WDYT?

### PR DESCRIPTION
**Pull request by hoechenberger**
Closes #12255

I also removed numerous unnecessary `# noqa: D102`'s from `__init__()` methods. (Unnecessary there if the class itself has a docstring)

### CODE REVIEW

File: mne/viz/_3d.py [hunk 1]
```diff
@@ -2187,8 +2187,9 @@ def link_brains(brains, time=True, camera=False, colorbar=True, picking=False):
 
     if _get_3d_backend() != "pyvistaqt":
         raise NotImplementedError(
-            "Expected 3d backend is pyvistaqt but"
-            " {} was given.".format(_get_3d_backend())
+            "Expected 3d backend is pyvistaqt but" " {} was given.".format(
```
Comment (cbrnr) [inline mne/viz/_3d.py:2190]: Things like this should be manually edited.

Comment (cbrnr) [reply mne/viz/_3d.py:2190]: Two strings on the same line?

Comment (cbrnr) [reply mne/viz/_3d.py:2190]: Oh my! It would be good to fix those. I get 442 results on main, so if necessary I'll manually check all of them.

**Comment by cbrnr** [comment]
@mscheltienne https://marketplace.visualstudio.com/items?itemName=charliermarsh.ruff

**Comment by mscheltienne** [comment]
OK found the linked issue, speed sounds good, although I would also be inclined to give them some time to refine this new functionality.
@cbrnr thanks,I knew about that one and have it setup, not sure if it's integrated in pycharm, and definitely not in spyder.

**Comment by larsoner** [comment]
> Style checks are still duplicated with "Tests / Style" (GHA) and "pre-commit.ci" checks. I think we can remove the "pre-commit.ci" check, since we disabled autofix. @larsoner?

I'd rather move more stuff to `pre-commit.ci` when possible actually. It tends to be faster than GHA and distributes load better to have it there. I think it makes sense (and we have been moving toward) to have pre-commit + pre-commit.ci take care of as much as possible, and leave the checks that are too slow for pre-commit (like `twine check`) to the remaining GHA style check

File: mne/beamformer/tests/test_rap_music.py [hunk 2]
```diff
@@ -50,9 +50,7 @@ def simu_data(evoked, forward, noise_cov, n_dipoles, times, nave=1):
     # Generate the two dipoles data
     mu, sigma = 0.1, 0.005
     s1 = (
-        1
-        / (sigma * np.sqrt(2 * np.pi))
-        * np.exp(-((times - mu) ** 2) / (2 * sigma**2))
+        1 / (sigma * np.sqrt(2 * np.pi)) * np.exp(-((times - mu) ** 2) / (2 * sigma**2))
```
Comment (larsoner) [inline mne/beamformer/tests/test_rap_music.py:53]: I was wondering why we had a ton of lines removed... I'm guessing it's mostly from this. I consider this a nice readability win

**Review by larsoner — COMMENTED** [review]
+1 for merge from me, with or without looking into the `" + ".*"` stuff. Overall I think this improves things and we can manually fix the `" + ".*"` when we come across them (or wait for the tool to get better and fix these for us automatically!).

**Comment by hoechenberger** [comment]
@larsoner Currenrly, some GHA jobs like the one that runs pytest depend on the "style" job, which runs the pre-commit hooks. This is to avoid running pytest as long as there are still code style issues. 

Can we make our GHA runs on pre-commit.ci return codes?

**Comment by hoechenberger** [comment]
But that would mean we run the pre-commit hooks twice (pre-commit.ci and GHA)? Unless we remove the job from GHA, businesses then the others cannot be (easily) conditional 

What do you suggest we do?

**Comment by larsoner** [comment]
I think the way things work right now are okay. There's always room for improvement but I think we have more important stuff to work on (diminishing returns tweaking CIs further)

File: mne/preprocessing/interpolate.py [hunk 3]
```diff
@@ -123,8 +123,9 @@ def interpolate_bridged_electrodes(inst, bridged_idx, bad_limit=4):
     pos = montage.get_positions()
     if pos["coord_frame"] != "head":
         raise RuntimeError(
-            "Montage channel positions must be in ``head``"
-            "got {}".format(pos["coord_frame"])
+            "Montage channel positions must be in ``head``" "got {}".format(
+                pos["coord_frame"]
+            )
```
Comment (drammock) [inline mne/preprocessing/interpolate.py:128]:
```suggestion
            f"Montage channel positions must be in ``head`` got {pos["coord_frame"]}"
```

File: mne/preprocessing/tests/test_annotate_amplitude.py [hunk 4]
```diff
@@ -247,11 +247,14 @@ def test_flat_bad_acq_skip():
...
-        "1741 1841 2011 2131 2141 2241 2531 2541 2611 2621".split()
-    ]
+    assert (
+        bads
+        == [  # MaxFilter finds the same 21 channels
+            "MEG%04d" % (int(num),)
+            for num in "141 331 421 431 611 641 1011 1021 1031 1241 1421 "
+            "1741 1841 2011 2131 2141 2241 2531 2541 2611 2621".split()
+        ]
+    )
```
Comment (drammock) [inline mne/preprocessing/tests/test_annotate_amplitude.py:257]: based on local testing, moving the comment to the end makes this get formatted much nicer
```suggestion
    assert bads == [
        f"MEG{num:04d}"
        for num in "141 331 421 431 611 641 1011 1021 1031 1241 1421 "
        "1741 1841 2011 2131 2141 2241 2531 2541 2611 2621".split()
    ]  # MaxFilter finds the same 21 channels
```

File: mne/tests/test_annotations.py [hunk 5]
```diff
@@ -425,7 +425,11 @@ def test_raw_reject(first_samp):
     with pytest.warns(RuntimeWarning, match="outside the data range"):
         raw.set_annotations(Annotations([2, 100, 105, 148], [2, 8, 5, 8], "BAD"))
     data, times = raw.get_data(
-        [0, 1, 3, 4], 100, 11200, "omit", return_times=True  # 1-112 s
+        [0, 1, 3, 4],
+        100,
+        11200,
+        "omit",
+        return_times=True,  # 1-112 s
     )
```
Comment (drammock) [inline mne/tests/test_annotations.py:433]: here again moving the comment yields better results

```suggestion
    data, times = raw.get_data(
        [0, 1, 3, 4], 100, 11200, "omit", return_times=True
    )  # 1-112 s
```

File: mne/viz/_3d.py [hunk 6]
```diff
@@ -2187,8 +2187,9 @@ def link_brains(brains, time=True, camera=False, colorbar=True, picking=False):
...
+                _get_3d_backend()
+            )
```
Comment (drammock) [inline mne/viz/_3d.py:2192]:
```suggestion
            f"Expected 3d backend is pyvistaqt but {_get_3d_backend()} was given."
```

File: mne/viz/epochs.py [hunk 7]
```diff
@@ -498,8 +499,9 @@ def _validate_fig_and_axes(fig, axes, group_by, evoked, colorbar, clear=False):
     if isinstance(axes, list):
         if len(axes) != n_axes:
             raise ValueError(
-                '{}"axes" must be length {}, got {}.'
-                "".format(prefix, n_axes, len(axes))
+                '{}"axes" must be length {}, got {}.' "".format(
+                    prefix, n_axes, len(axes)
+                )
```
Comment (drammock) [inline mne/viz/epochs.py:504]:
```suggestion
                f'{prefix}"axes" must be length {n_axes}, got {len(axes)}.'
```

File: mne/beamformer/tests/test_rap_music.py (hunk 2 above)
Comment (hoechenberger) [reply mne/beamformer/tests/test_rap_music.py:53]: Same.

File: mne/forward/tests/test_field_interpolation.py [hunk 8]
```diff
@@ -237,10 +237,16 @@ def test_make_field_map_meeg():
         assert_allclose(map_["data"].min(), min_, rtol=5e-2)
     # calculated from correct looking mapping on 2015/12/26
     assert_allclose(
-        np.sqrt(np.sum(maps[0]["data"] ** 2)), 19.0903, atol=1e-3, rtol=1e-3  # 16.6088,
+        np.sqrt(np.sum(maps[0]["data"] ** 2)),
+        19.0903,
+        atol=1e-3,
+        rtol=1e-3,  # 16.6088,
```
Comment (hoechenberger) [inline mne/forward/tests/test_field_interpolation.py:243]:
```suggestion
        rtol=1e-3,
```

File: mne/forward/tests/test_field_interpolation.py [hunk 9]
```diff
@@ -237,10 +237,16 @@ def test_make_field_map_meeg():
...
     )
     assert_allclose(
-        np.sqrt(np.sum(maps[1]["data"] ** 2)), 19.4748, atol=1e-3, rtol=1e-3  # 20.1245,
+        np.sqrt(np.sum(maps[1]["data"] ** 2)),
+        19.4748,
+        atol=1e-3,
+        rtol=1e-3,  # 20.1245,
```
Comment (hoechenberger) [inline mne/forward/tests/test_field_interpolation.py:249]:
```suggestion
        rtol=1e-3,
```

File: mne/inverse_sparse/mxne_optim.py [hunk 10]
```diff
@@ -977,9 +977,9 @@ def norm_epsilon(Y, l1_ratio, phi, w_space=1.0, w_time=None):
         p_sum_w2 = np.cumsum(w_time**2)
         p_sum_Yw = np.cumsum(Y * w_time)
         upper = p_sum_Y2 / (Y / w_time) ** 2 - 2.0 * p_sum_Yw / (Y / w_time) + p_sum_w2
-    upper_greater = np.where(
-        upper > w_space**2 * (1.0 - l1_ratio) ** 2 / l1_ratio**2
-    )[0]
+    upper_greater = np.where(upper > w_space**2 * (1.0 - l1_ratio) ** 2 / l1_ratio**2)[
+        0
+    ]
```
[the remark this answers is not in the document]
Comment (hoechenberger) [reply mne/inverse_sparse/mxne_optim.py:982]: None that I know of. Seems to be a rare edge case … is this one a blocker for you?

File: tutorials/stats-source-space/60_cluster_rmANOVA_spatiotemporal.py [hunk 11]
```diff
@@ -285,9 +285,7 @@ def stat_fun(*args):
 
 inds_t, inds_v = [
     (clusters[cluster_ind]) for ii, cluster_ind in enumerate(good_cluster_inds)
-][
-    0
-]  # first cluster
```
Comment (hoechenberger) [inline tutorials/stats-source-space/60_cluster_rmANOVA_spatiotemporal.py:290]: @drammock here's an example where things are reversed and are looking better with ruff

File: mne/viz/topomap.py [hunk 12]
```diff
@@ -1220,7 +1220,8 @@ def _plot_topomap(
         elif len(pos["chs"]) != data.shape[0]:
             raise ValueError(
                 "Number of channels in the Info object (%s) and "
-                "the data array (%s) do not match. " % (len(pos["chs"]), data.shape[0])
+                "the data array (%s) do not match. "
+                % (len(pos["chs"]), data.shape[0])
                 + info_help
```
[the remark this answers is not in the document]
Comment (hoechenberger) [reply mne/viz/topomap.py:1225]: this suggestions cannot be applied, GH claims the lines have been deleted

File: mne/inverse_sparse/mxne_optim.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (drammock) [reply mne/inverse_sparse/mxne_optim.py:982]: Not a blocker. There was at least one case elsewhere in the diff where black was doing this and ruff fixed it 🤷🏻

File: mne/viz/topomap.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (drammock) [reply mne/viz/topomap.py:1224]: Whatever, GH. I can push the change later if you want. But there are likely still a bunch of these to fix that happened to not be part of the diff so far, so ok to just leave it, too, and address it later

[the remark this answers is not in the document]
Comment (hoechenberger) [reply mne/viz/topomap.py]: I resolved this manually

**Comment by hoechenberger** [comment]
I think this is good to go.

**Comment by hoechenberger** [comment]
@mscheltienne Okay with you if we move forward here?

**Comment by mscheltienne** [comment]
Sure!

**Comment by larsoner** [comment]
Okay let's give it a shot!

I'll add e7dd1588013179013a50d3f6b8e8f9ae0a185783 to `.git-blame-revs-ignore` in some follow-up PR

**Pull request merged**